pkg/reconcile/pipeline/context/impl.go

```diff
@@ -120,51 +120,61 @@ func (i *impl) Services() ([]pipeline.Service, error) {
 	return services, nil
 }
 
+// Application return a list of applications.
+// And if no application found, return an error
 func (i *impl) Applications() ([]pipeline.Application, error) {
 	if i.applications == nil {
-		if ref := i.serviceBinding.Spec.Application; ref != nil {
-			gvr, err := i.typeLookup.ResourceForReferable(ref)
+		ref := i.serviceBinding.Spec.Application
+		gvr, err := i.typeLookup.ResourceForReferable(&ref)
+		if err != nil {
+			return nil, err
+		}
+		if i.serviceBinding.Spec.Application.Name != "" {
+			u, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).Get(context.Background(), ref.Name, metav1.GetOptions{})
 			if err != nil {
-				return nil, err
+				return nil, emptyApplicationsErr{err}
+			}
+			i.applications = append(i.applications, &application{gvr: gvr, persistedResource: u, bindingPath: i.serviceBinding.Spec.Application.BindingPath})
+		}
+		if i.serviceBinding.Spec.Application.LabelSelector != nil && i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+			matchLabels := i.serviceBinding.Spec.Application.LabelSelector.MatchLabels
+			opts := metav1.ListOptions{
+				LabelSelector: labels.Set(matchLabels).String(),
+			}
+
+			objList, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).List(context.Background(), opts)
+			if err != nil {
+				return nil, emptyApplicationsErr{err}
```

Comment:
we do not know if application list is empty or not, we should just pass through the obtained error to the caller:

```
return nil, err
```

Reply:
Done. 34cd1e9bf29f243191f69e5352b0440b288b3927